containers/datalab/content/run-with-gce.sh

```diff
@@ -0,0 +1,145 @@
+#!/bin/bash -e
+
+# Copyright 2016 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+USAGE='USAGE: One of...
+
+To run locally:
+
+    docker run -it -p "8081:8080" -v "${HOME}:/content" gcr.io/cloud-datalab/datalab
+
+Or, to connect to a kernel gateway in a GCE VM;
+
+    docker run -it -p "8081:8080" -v "${HOME}:/content" \
+      -e "GATEWAY_VM=${PROJECT_ID}/${ZONE}/${INSTANCE}"
+      gcr.io/cloud-datalab/datalab
+'
+
+ERR_MALFORMED_GATEWAY=1
+ERR_LOGIN=2
+ERR_PROJECT_NOT_FOUND=3
+ERR_ZONE_NOT_FOUND=4
+ERR_INSTANCE_NOT_FOUND=5
+ERR_DEPLOY=6
+ERR_TUNNEL_FAILED=7
+ERR_GATEWAY_FAILED=7
+
+# Run the 'setup-env' script to ensure that gcloud has been told to use
+# a config directory under the mounted volume (so that the results of
+# 'gcloud auth login' are persisted).
+source /datalab/setup-env.sh
+export HOME=/content
+
+export PROJECT_ID=${PROJECT_ID:-`gcloud config list -q --format 'value(core.project)' 2> /dev/null`}
+export ZONE=${ZONE:-`gcloud config list -q --format 'value(compute.zone)' 2> /dev/null`}
+
+if [[ -n "${GATEWAY_VM}" ]]; then
+  GATEWAY_PART_1=`echo "${GATEWAY_VM}" | cut -d '/' -f 1`
+  GATEWAY_PART_2=`echo "${GATEWAY_VM}" | cut -d '/' -f 2`
+  GATEWAY_PART_3=`echo "${GATEWAY_VM}" | cut -d '/' -f 3`
+
+  if [[ -z "${GATEWAY_PART_3}" &&  -z "${GATEWAY_PART_3}" ]]; then
+    export INSTANCE="${GATEWAY_PART_1}"
+  elif [[ -z "${GATEWAY_PART_1}" || -z "${GATEWAY_PART_2}" || -z "${GATEWAY_PART_3}" ]]; then
+    echo "Malformed gateway VM name"
+    echo "${USAGE}"
+    exit "${ERR_MALFORMED_GATEWAY}"
+  else
+    export PROJECT_ID="${GATEWAY_PART_1}"
+    export ZONE="${GATEWAY_PART_2}"
+    export INSTANCE="${GATEWAY_PART_3}"
+  fi
+fi
+
+if [[ -n "${INSTANCE}" ]]; then
+  USER_EMAIL=`gcloud info --format="value(config.account)"`
+  if [[ -z "${USER_EMAIL}" ]]; then
+    FAILED_LOGIN=""
+    gcloud auth login || FAILED_LOGIN="true"
+    if [[ -n "${FAILED_LOGIN}" ]]; then
+      echo "Failed to log in to gcloud"
+      exit "${ERR_LOGIN}"
```

Comment:
The only error cases here that might reasonably be transient are the ones where we try to verify the gateway is usable. I have added messages to those cases indicating that a retry may work.

However, for every other one simply rerunning the command shouldn't work unless the user also changes something (such as not cancelling the auth flow). As such, I don't think asking them to retry is going to be helpful.